### DISCUSSION TITLE
Clarify that Dagger behaves more like `docker build` instead of like `docker run`

### DISCUSSION THREAD

**Discussion by thecooldrop** [dagger/dagger#12171] [Help]
### What is the issue?

I have recently started using Dagger and have experienced a lot of friction, due to not being able to understand how Dagger behaves. Here are two examples of things, which I found confusing as user:

- When running `with_exec` on container, and starting a process in background such as socat, for example using `.with_exec(["/bin/sh", "-c", "socat tcp-listen:2375,reuseaddr, fork tcp:dockerd:2375 &"]).with_exec(["ps", "aux"])`, then `ps aux` does not display the socat process, because it was terminated.
- When building Dockerfiles using the build command on container, it is not clear that the Dockerfile steps are executed in context of the container from which the build was started, and not in separate container. This means that the mounted services are accessible via IP in the Dockerfile ( see the discussion starting here: https://discord.com/channels/707636530424053791/1236088075507994664/1322377501569454181 and nicely explained by marcosnils )
- It is weird that mounted Services are accessible via IP in the Dockerfile builds, but not via hostname
- It is weird that `host.docker.internal` is not accessible within Dockerfiles.

Both of these "corner cases" would be a lot clearer to me if the documentation was clear on following points:

- The fact that Dagger acts more like Docker build, and not like Docker run
- Clarity about what is accessible where. It is currently not intuitive what is shared between a) Dagger engine and containers started within the engine b) between containers which are started by Dagger engine. For example, for me it is not at all intuitive that bound Services are accessible during build steps within Dockerfile via IP, but the files from within the host container are not accessible.

To sum up, this post is about:

- Improving documentation regarding the fact that Dagger acts more like Docker build, and not like Docker run.
- Improving documentation on execution environment within Dagger Engine. Here it should be documented what aspects of containers are shared, and which are not shared, so that clearer picture is available to users.

**Comment by jpadams** [comment]
Thank you @thecooldrop. Agree that we can improve the docs in these areas.
- Have you seen any docs that do a great job with setting that kind of context for new users?
- Are there existing sections in the Dagger docs where you'd expect to find the information you were looking for? Which ones?
- Or do we need a new section or sections? If  so, what would you call them?

**Comment by thecooldrop** [comment]
Regarding your questions:

- I have no reference to an example from other projects, which could serve as a guide toward this. Gradle may be good, since it also starts out by demonstrating a simple usecase, and then iteratively deepens the discussion of specific areas in additional sections. For me usually structuring the documentation around basic usecase and then providing a deeper look, in one colocated area is usually very helpful.
- There is currently no section which I would put this in, but I would expect this section to be linked from many other sections, and to link to other related sections. There should at least be links from: cookbook, services sections.
- I think we need two new sections at least. The first sections should be "Concepts" sections, providing an overview of the concepts which are relevant for working with Dagger and how they are related. Currently we have section for each "concept" (such as services, secrets, volumes and etc...), but section about "execution environment" describing the environment in which the functions execute is currently not there. My expectation would be that the "Execution environment" section describes what is accessible for our functions, and what additional containers mean when spawned by functions within Dagger Engine. What volumes are accessible from where, what services, what DNS names, what files...etc?
